File: src/main/java/com/fishercoder/solutions/_698.java
package com.fishercoder.solutions;

import java.util.Arrays;

public class _698 {

    public static class Solution1 {
        public boolean canPartitionKSubsets(int[] nums, int k) {
            long sum = 0;
            for (int num : nums) {
                sum += num;
            }
            if (sum % k != 0) {
                return false;
            }
            int equalSum = (int) (sum / k);
            boolean[] visited = new boolean[nums.length];
            return canPartition(nums, visited, 0, k, 0, 0, equalSum);
        }

        private boolean canPartition(int[] nums, boolean[] visited, int startIndex, int k, int currSum, int currNum, int target) {
            if (k == 1) {
                return true;
            }
            if (currSum == target && currNum > 0) {
                /**Everytime when we get currSum == target, we'll start from index 0 and look up the numbers that are not used yet
                 * and try to find another sum that could equal to target*/
                return canPartition(nums, visited, 0, k - 1, 0, 0, target);
            }
            for (int i = startIndex; i < nums.length; i++) {
                if (!visited[i]) {
                    visited[i] = true;
                    if (canPartition(nums, visited, i + 1, k, currSum + nums[i], currNum++, target)) {
                        return true;
                    }
                    visited[i] = false;
                }
            }
            return false;
        }
    }

    public static class Solution2 {
        /**
         * I'm glad that I figured out below solution completely on my own on 9/30/2021.
         * Backtracking is so beautiful!
         * <p>
         * Although not super concise, and I've added a sorting step, it's completely original.
         */
        public boolean canPartitionKSubsets(int[] nums, int k) {
            Arrays.sort(nums);
            long sum = 0l;
            for (int num : nums) {
                sum += num;
            }
            int ave = (int) sum / k;
            if (sum % k != 0) {
                return false;
            }
            boolean[] used = new boolean[nums.length];
            int found = 0;
            for (int i = nums.length - 1; i >= 0; i--) {
                if (!used[i]) {
                    used[i] = true;
                    found += recursive(nums, used, ave, nums[i], i);
                }
            }
            return k == found;
        }

        private int recursive(int[] nums, boolean[] used, int targetSum, int currSum, int currIndex) {
            if (currSum == targetSum) {
                return 1;
            } else if (currSum > targetSum) {
                used[currIndex] = false;
                return 0;
            } else {
                if (currIndex > 0) {
                    for (int i = currIndex; i > 0; i--) {
                        if (!used[i - 1]) {
                            used[i - 1] = true;
                            int found = recursive(nums, used, targetSum, currSum + nums[i - 1], i - 1);
                            if (found == 1) {
                                return found;
                            }
                            used[i - 1] = false;//this is the backtracking step: reset this number to be available if not found
                        }
                    }
                }
                return 0;
            }
        }
    }

}

File: cpp/_698.cpp
//Partition to K Equal Sum Subsets

class Solution {
public:
    
    bool dfs(vector<int> &nums, vector<int> visited, int idx,int k,int currSum,int targetSum){
        if(k==1){
            //All subsets are found
            return true;
        }
        if(currSum==targetSum){
            //one subset found
            return dfs(nums,visited,0,k-1,0,targetSum);
        }
        for(int i=idx;i<nums.size();i++){
            if(!visited[i]){
                visited[i]=true;
                if(dfs(nums,visited,i+1,k,currSum+nums[i],targetSum)){
                    return true;
                }
                visited[i]=false;
            }
        }
        return false;
    }
    bool canPartitionKSubsets(vector<int>& nums, int k) {
        vector<int> visited(nums.size(),false);
        int sum=0;
        for(int i=0;i<nums.size();i++){
            sum += nums[i];
        }
        if(sum%k!=0){
            return false;
        }
        return dfs(nums,visited,0,k,0,sum/k);
    }
};
